### ISSUE TITLE
Question about multiple updates for a SSBO or UBO

### ISSUE THREAD

**Issue by jaebaek** [Igalia/vkrunner#46]
Based on vkrunner spec, there is no limitation on updating one SSBO or UBO multiple times e.g.,
```
ssbo 0:0 subdata vec4 1 2 3 4
ssbo 0:0 subdata vec4 1 2 3 4 5 6 7 8 9 10 11 12
```

I am curious what is expected behavior for this case.

**Comment by jaebaek** [comment]
In particular, I am curious the result of [this script](https://github.com/google/amber/blob/master/tests/cases/multiple_ssbo_update_with_graphics_pipeline.amber) should be a pass or failure.

**Comment by dj2** [comment]
The result of the above would be an ssbo with `1 2 3 4 5 6 7 8 9 10 11 12` in it. The second call would increase the size and overwrite all the data. The ssbo is sized with the largest ssbo call that is seen.

**Comment by bpeel** [comment]
Putting multiple values in an `ssbo subdata` command is intended to be able to update an array of values. So for example this line:

```
ssbo 0 subdata vec4 0     1.0 2.0 3.0 4.0 5.0 6.0 7.0 8.0 9.0 10.0 11.0 12.0
```

could be used to set the values in an SSBO declared like this:

```GLSL
layout(binding = 0, std140) buffer block {
        vec4 my_favourite_vectors[3];
}
```

Note however that VkRunner always assumes std140 layout. The default in GLSL for SSBOs is std430 so in your example VkRunner will put all the values in the wrong offsets and presumably the test won’t work. The plan is to eventually address issue #32 and have a way to tell VkRunner what buffer layout to use when updating buffers. If we do this it might be nice to make it default to std430 for SSBO updates. Whatever we do it would be great if we could try to get Amber and VkRunner to behave the same way.

**Comment by bpeel** [comment]
I’ve posted a pull request (#47) to make VkRunner default to std430 for SSBOs and push constants. This makes the script mentioned [above](https://github.com/Igalia/vkrunner/issues/46#issuecomment-448358190) work.

**Comment by bpeel** [comment]
I now realise I completely misunderstood your question :)

The comment from dj2 is correct, updating multiple times just overwrites any previous updates.

However note there is a subtlety here in that all drawing and compute commands are just added to the command buffer and buffer writes are implemented by just writing to the mapped memory. For example:

```
ssbo 0 subdata float 0     8.0
compute 1 1 1
ssbo 0 subdata float 0     42.0
compute 1 1 1
```

In this case both compute commands will end up seeing the value 42 because VkRunner doesn’t do anything to try to flush the command buffer between buffer updates. This is mentioned in the README.

However this is irrelevant to the example in the question.

**Comment by dj2** [comment]
For the `ssbo, compute, ssbo, compute` case, Amber will end up differing from VkRunner as the `compute` commands will only use the `ssbo` data declared before the command. So, the first `compute` would use 8 and the second would use 42 which I think is closer to what you'd expect from reading the test file.

**Comment by jaebaek** [comment]
@bpeel Thank you for clarifying the status of VkRunner.

In terms of the memory layout issue, I considered it for a while and realized we should separate the actual content in buffer from the memory layout.
Whatever GLSL shader uses for layouts, we fill the buffer simply with given data.

For example, for both of
```
layout(set = 0, binding = 0) buffer block {
  float a[3];   // Offset: 0, array stride:   4
};
```
and
```
layout(set = 0, binding = 0, std140) buffer block {
  float a[3];   // Offset: 0, array stride:   16
};
```

user can create a buffer with
```
ssbo 0:0 subdata float  0  1  0  0  0 \
                           2  0  0  0 \
                           3
```

For the first layout, `float a[3]` will be `{1, 0, 0}`.
For the second layout, `float a[3]` will be `{1, 2, 3}`.

I think this would be more intuitive than specifying memory layout.

@bpeel what do you think?

**Comment by jaebaek** [comment]
In terms of memory overwrite .. I want to consider what would be better design more.
I guess there will be probably more weird cases.
I will add more comments for it after thinking of it.

**Comment by dj2** [comment]
There is not much to be done about the overwrite. VkRunner already works that way and what it does makes sense, if you request putting that data in the ssbo at the same location it makes sense that it overwrites.

**Comment by bpeel** [comment]
@jaebaek I agree the suggestion about just always tightly packing is simpler to understand but I’m not sure if it’s more convenient. If we make setting UBOs default to std140 and SSBOs default to std430 then most of the time the user doesn’t need to even think about it and it would just do the right thing to set the array values without the user having to understand the layouts. With your suggestion the user has to understand the padding and manually add it.

What do you think about setting the values for matrix types? If we keep the ability to set a matrix type with a single command then we can’t avoid having to know the layout.

Eg, with a command like this:

```
ssbo 0 subdata mat2 0  1 2 3 4
```

If the layout is std430 then the floats would just be tightly packed in memory like this:

    1 2 3 4

otherwise each column would be padded to a vec4 like this:

    1 2 0 0 3 4

Even with std430 there ends up being padding between the columns if the number of rows is 3, so we can’t just say “everything is always tightly packed” and ignore std140.

**Comment by jaebaek** [comment]
I understand what you mean but I am worried that users would be confused when reading some scripts and thought the behavior is unexpected. In addition, SPIR-V assembly shows the layout accurately. Matching the explicit memory layout shown it SPIR-V to the actual content buffer would be more clear.

I want to discuss this issue in Vulkan CTS meeting and take opinions of other people.
I also want to discuss the multiple updates for a SSBO as well.
Please join the Vulkan CTS meeting tomorrow if you are available.

**Comment by jaebaek** [comment]
Before discussing it in the meeting, I want to check your design with one example.

If you or users intend using `layout std140` command for `std140` in GLSL, how do you separate following two cases?

```
layout (set = 0, binding = 0, std140) buffer block {
  float a[3];   // array stride = 16
};
```

and

```
layout (set = 0, binding = 0, std140) buffer block {
  float a, b, c;   // stride = 4
};
```

When we want to fill `1, 2, 3`, we should use
```
layout std140
ssbo 0 subdata 0 1 2 3
```
for the first GLSL shader.

For the second GLSL shader, we must use
```
layout std430
ssbo 0 subdata 0 1 2 3
```

I think there are more weird cases and just setting up buffers without specific layout is much more clear.

**Comment by jaebaek** [comment]
In terms of multiple updates for a SSBO, I think we should apply updates directly as @dj2 mentioned.

For example,

```
ssbo 0 subdata    0    8.0
compute 1 1 1    # use 8.0 for SSBO at 0:0

ssbo 0 subdata    0   48.0
compute 1 1 1    # use 48.0 for SSBO at 0:0

ssbo 0 subdata 1024    1   2 3 4 5 6 7 8 9 10 11 ... 4096
compute 1 1 1    # increase the size of SSBO at 0:0 and use 48, 0, 0, ..., 0, 1, 2, ..., 4096
```

I think this is more intuitive to users.

**Comment by dj2** [comment]
I'm not sure I understand why you mean by setting up buffers with explicit layouts? The buffer needs a layout so the data is understandable. Are you suggesting that we try to extract the layout from the shader? Or assume the user always used a layout that matches the shader and can just copy the data directly?

I think having a documented default (ssbo, ubo are 430 and push is 140 or whatever the defaults from glslang are) is fine and if you want something different you'd do the ssbo layout 140 command. You only need to do the layout command if it differs from the defaults.

**Comment by bpeel** [comment]
I think in the second case it seems reasonable just to use separate subdata commands with the explicit offsets like this:

```
ssbo 0 subdata float 0 1
ssbo 0 subdata float 4 2
ssbo 0 subdata float 8 3
```

Just packing the data doesn’t help to set all of the data in all cases. For example if there were different types in the buffer you would need separate commands anyway:

```GLSL
layout(binding = 0) buffer block {
        double a;
        float b;
        int c;
};
```

```
ssbo 0 subdata double 0 1
ssbo 0 subdata float 8 2
ssbo 0 subdata int 12 3
```

**Comment by bpeel** [comment]
Just to clarify, the layout commands in the [test] section and specifying multiple values in a single ssbo subdata command are intended to just work as a convenience for making it easier to set values. If you don’t want to use it or the shader is using a weird layout, you can always set individual values with multiple ssbo subdata commands and explicit offsets. Eg, with `ssbo subdata float <offset> <value>` you can put a float at any offset in the buffer. There is nothing stopping you using multiple commands and setting the data is very cheap. It just amounts to a memcpy into the permanently mapped buffer.

**Comment by jaebaek** [comment]
Ok. If you want to design VkRunner script like that way, we will follow your opinions. Could you please describe what are expected behavior of `ssbo` commands after turning on `layout std140` command in document?

e.g.,
```
layout std140
ssbo 0 subdata float 0 1 2 3
ssbo 0 subdata vec3 0 1 2 3
ssbo 0 subdata vec3 0 1 2 3 4 5 6 7 8 9
ssbo 0 subdata mat3x3 0 1 2 3 4 5 6 7 8 9
```

**Comment by bpeel** [comment]
Ok, I have written a more detailed description of how it’s meant to work in the proposed layout branch [here](https://github.com/Igalia/vkrunner/tree/nroberts/layouts#setting-buffer-subdata).

**Comment by jaebaek** [comment]
In addition, could you please add a description for expected behaviors of multiple updates for a SSBO?

In the following example, does VkRunner use only the final SSBO for all compute?

```
ssbo 0 subdata float 0 1.0
compute 1 1 1
ssbo 0 subdata float 0 2.0
compute 1 1 1
ssbo 0 subdata float 36 1.0 2.0 3.0 ... ( more than 4000 numbers) ... 4096.0
compute 1 1 1
probe ssbo float 0 ~= 2.0
ssbo 0 subdata float 4000 1.0 2.0 3.0 ... ( more than 4000 numbers) ... 4096.0
compute 1 1 1
probe ssbo float 4000 ~= 1.0
```

i.e., is it the same with the following?

```
ssbo 0 subdata float 0 1.0
ssbo 0 subdata float 0 2.0
ssbo 0 subdata float 36 1.0 2.0 3.0 ... ( more than 4000 numbers) ... 4096.0
ssbo 0 subdata float 4000 1.0 2.0 3.0 ... ( more than 4000 numbers) ... 4096.0

compute 1 1 1
compute 1 1 1
compute 1 1 1
probe ssbo float 0 ~= 2.0
compute 1 1 1
probe ssbo float 4000 ~= 1.0
```

**Comment by bpeel** [comment]
That is not quite how it works, no.

Any draw or compute command is just added to a command buffer. Whenever a probe command is reached or the script finishes the command buffer is ended, submitted and VkRunner waits for it to complete.

Any buffer update command just directly writes into the memory-mapped memory without considering the command buffer.

That means that something like the following will probably do as you expect:

```
[compute shader]
#version 450
layout(binding = 0) uniform ubo { float square; };
layout(binding = 1) buffer block { float square_root; };
void
main()
{
        square_root = sqrt(square);
}

[test]
ssbo 1 4

# Set the input to 16
uniform ubo 0 float 0 16.0
# Compute with that
compute 1 1 1
# Check that it worked
probe ssbo float 1 0 ~= 4.0

# Do it again with a different input
uniform ubo 0 float 0 4.0
compute 1 1 1
probe ssbo float 1 0 ~= 2.0
```

The confusion arises if you update multiple times without a probe command:

```
uniform ubo 0 float 0 16.0
compute 1 1 1
uniform ubo 0 float 0 4.0
compute 1 1 1
# The command buffer is only submitted here so both compute commands will see 4.0
probe ssbo float 1 0 ~= 2.0
```

It would be a fairly trivial change to make VkRunner also flush any pending command buffer whenever there is a buffer update and then I guess it would work like Amber. However there is a slight snag to this if you mix push constants with buffer updates. This is because push constants are also just commands in the command buffer and their value is not persistent across command buffer submissions. For example, the following script wouldn’t work if we change it to use that behaviour:

```
# Set a push constant
uniform 0 16.0
# Try to set a ubo value. This would end up flushing the command buffer
# and the push constant value would be forgotten.
uniform ubo 0 float 0 25.0
```

I guess we could try to do something to rearrange the commands so that it would work but I think that might be getting too far into the territory of doing too much unpredictable magic. For that reason I think it needs to be quite predictable when the script is going to flush the command buffer or not.

**Comment by dneto0** [comment]
> I guess we could try to do something to rearrange the commands so that it would work but I think that might be getting too far into the territory of doing too much unpredictable magic.

Agreed.  We mainly want to understand the expected semantics. It's the unexpected mixing of commands that needed he better explanation.

Reading the doc very carefully, I see the key is the part starting with "Note that the buffer is just updated by writing into a memory mapped view of it..." is in the area describing the "uniform" command, and that it carries over to ssbo because ssbo is like 'uniform ubo' except for the buffer type.

It might help to have a high level description at the beginning describing the distinction between creating buffers and other inputs and _scheduling_ dispatches and draws (schedule via recording a command), vs actually flushing the commands out for execution at when the next probe is encountered.

**Comment by dneto0** [comment]
For the case where there are multiple dispatches without a probe between:

```
uniform ubo 0 float 0 16.0
compute 1 1 1
uniform ubo 0 float 0 4.0
compute 1 1 1
# The command buffer is only submitted here so both compute commands will see 4.0
probe ssbo float 1 0 ~= 2.0
```
If the user truly wants to run two dispatches and use different values for them, then perhaps we can invent a "flush" command which is the same as a "probe" with a trivially satisfied check.  That is, the flush will send the data to the GPU, flush the commands, and get the buffer data back:

```
uniform ubo 0 float 0 16.0
compute 1 1 1

# The flush will cause the compute dispatch to use value 16.0, and then map the
# buffers back again.   Don't care about checking any output values.

flush

# Start a new command sequence, but can overwrite buffer data since it
# again is mapped into host space.

uniform ubo 0 float 0 4.0
compute 1 1 1

# The command buffer is submitted and the dispatch will see value 4.0

probe ssbo float 1 0 ~= 2.0
```

Is there utility in the pattern? Maybe, because I only want to check the final result.

**Comment by bpeel** [comment]
I like the idea of the flush command. And yes, it would be good to add a section to the documentation to explicitly specify when the flushes take place. We could say that it only happens for a probe command, the flush command or the end of the script.

Piglit has quite a few tests for GL that use this pattern already. For example, if it wants to test the sqrt function in a fragment shader, it might draw several little rectangles at different locations each with different uniform inputs and then only probe the whole framebuffer at the end to make sure all of the rectangles are green. In this particular case however it would probably be more convenient to use push constants.

**Comment by dj2** [comment]
Trying to catchup after the holiday, please let me know if I'm off track here.

There are two different things being discussed in this issue:

 1- Layout of data in SSBOs and how that appears in the script
 2- Multiple SSBO updates and how they interact with compute/draw calls.

1- Layout of data in SSBOs and how that appears in the script
For the layout of data case, my understanding is, we have to deal with std430
and std140 in the source shader. At the script level we don't know what was
specified in the shader, so we don't know how to pack the data. This is fine
for individually specified items, (like float a, b, c in the shader) but doesn't
work for std140 with arrays (like float a[3]) because that array has 16 byte
alignment for each element of the array.

We can add the `layout std430` and `layout 140` commands to the parser to
specify what the layout should be (otherwise we default to ...?). This will work
for most cases, but given an ssbo with `1 2 3` in it, we have to know if the shader
is `float a, b, c` or `float a[3]`. It looks like the data type for the ssbo
doesn't encode information about if this is an array or not. Could that be added
to the data type definition then we'd know how much padding to add if needed?

2- Multiple SSBO updates and how they interact with compute/draw calls.
For the SSBO update case, with VkRunner currently all SSBO calls until a Probe
will be used for any compute or draw calls before that probe. This seems like
it could be the cause of confusing when reading a script file as something like:

```
ssbo 0 subdata float 0 1
compute
ssbo 0 subdata float 0 2
compute
probe
```

Wouldn't do what I expect, which is to use `1` for the first compute call and
`2` in the second compute call. We think the semantics in this case should be
along the lines of, when the `compute` is seen, the command buffer is flushed
and executed, the ssbo data is read back from the device to the host immediately
and stored on the host. The second ssbo call would then update that data, the
second `compute` sends the new data and reads back, the probe just uses the read
back data.

For push constants we'd need to send the buffer in each command submission if they
aren't stored on device.

The above is how Amber will work, in the case where VkRunner maintains the old
behaviour of sending all SSBO calls before a probe we'll probably flag that as
a parse error in the script and ask the user to re-order the ssob/compute/probe
calls so that behaviour isn't required.

**Comment by bpeel** [comment]
The decision of whether or not updating a buffer should cause the command buffer to flush also affects issue #52 where we want to add barriers for accessing SSBOs. It is much easier to implement this if updating a buffer is also a flush point.

Seeing as Amber seems to have already taken this decision there are now two good reasons to make VkRunner do the same. I’ve implemented this in the first two patches of PR #56. The second patch adds an explanation about flushing in the README.

**Comment by samuelig** [comment]
Repository moved to https://gitlab.freedesktop.org/mesa/vkrunner. Please re-submit this pull request there.
